File: 29Items-Projects/7Algorithms/8-COMPUTATIONAL-GEOMETRY/src/java/src/main/java/org/computationalgeometry/algorithms/PointInPolygon.java
package org.computationalgeometry.algorithms;

import org.computationalgeometry.model.Point2D;
import org.computationalgeometry.model.Polygon2D;
import org.computationalgeometry.util.NumericPredicates;

public final class PointInPolygon {
    private PointInPolygon() {
    }

    public enum Location {
        INSIDE,
        OUTSIDE,
        BOUNDARY
    }

    public static Location locate(Polygon2D polygon, Point2D point) {
        boolean inside = false;
        var vertices = polygon.vertices();

        for (int i = 0, j = vertices.size() - 1; i < vertices.size(); j = i++) {
            Point2D a = vertices.get(j);
            Point2D b = vertices.get(i);
            if (NumericPredicates.onSegment(a, b, point, NumericPredicates.DEFAULT_EPSILON)) {
                return Location.BOUNDARY;
            }

            boolean crosses = (a.y() > point.y()) != (b.y() > point.y());
            if (crosses) {
                double xAtY = (b.x() - a.x()) * (point.y() - a.y()) / (b.y() - a.y()) + a.x();
                if (point.x() < xAtY) {
                    inside = !inside;
                }
            }
        }

        return inside ? Location.INSIDE : Location.OUTSIDE;
    }
}

File: 29Items-Projects/7Algorithms/8-COMPUTATIONAL-GEOMETRY/src/python/geometry/algorithms/point_in_polygon.py
from enum import Enum

from geometry.model.point import Point2D
from geometry.model.polygon import Polygon2D
from geometry.util.numeric import DEFAULT_EPSILON, on_segment


class Location(str, Enum):
    INSIDE = "inside"
    OUTSIDE = "outside"
    BOUNDARY = "boundary"


def locate_point(polygon: Polygon2D, point: Point2D) -> Location:
    inside = False
    vertices = polygon.vertices

    for index, b in enumerate(vertices):
        a = vertices[index - 1]
        if on_segment(a, b, point, DEFAULT_EPSILON):
            return Location.BOUNDARY

        crosses = (a.y > point.y) != (b.y > point.y)
        if crosses:
            x_at_y = (b.x - a.x) * (point.y - a.y) / (b.y - a.y) + a.x
            if point.x < x_at_y:
                inside = not inside

    return Location.INSIDE if inside else Location.OUTSIDE
